tests/core/test_state_sync.py

```diff
@@ -465,19 +465,21 @@ def test_promote_snapshots(state_sync: EngineAdapterStateSync, make_snapshot: t.
 
     promotion_result = promote_snapshots(state_sync, [snapshot_a, snapshot_b_old], "prod")
 
-    assert set(promotion_result.added) == set([snapshot_a.table_info, snapshot_b_old.table_info])
+    assert sorted([x.json() for x in promotion_result.added]) == sorted(
```

Comment:
why did you need to change this test?

Reply:
It seems related to Pydantic version. With Pydantic 1 and I _think_ some Pydantic 2 versions the old version passes but when upgraded to latest it says they are different. So I serialize/sort the results and they correctly determine if they are equal.

Reply:
can you take a look at the actual pydantic error and let’s fix equality? this may patch the test but will still fail if we use eq check in regular code. id rather fix the equality in general as opposed to working around it in tests

Reply:
The issue is that Pydantic includes private attributes in the equality check:
https://github.com/pydantic/pydantic/blob/b785d5b1db4eccb922dd2bf17faea5fc842e6a57/pydantic/main.py#L880

We use these for caching values. In this case it is the fully qualified table: https://github.com/TobikoData/sqlmesh/blob/d5fd1b23a8cbb6c0eca1313af65762ebf09c006c/sqlmesh/core/snapshot/definition.py#L371-L375

This is actually a project wide issue that we don't currently address. A bad fix, just to demonstrate the issue, would be this:
```
    def __eq__(self, other):
        return (self.copy().__pydantic_private__ or {}).clear().__eq__((other.copy().__pydantic_private__ or {}).clear()) 
```

The problem obviously being that we would create copies on each equality check. It would be good to sync with @izeigerman on some options since we knows Pydantic the best. I'm fine with fixing this issue being out of scope of this PR.

Reply:
i don't really agree with that given you've already altered the scope of this PR by changing the tests to json(), i think we should fix it, in snapshot for example, we already over equality to simply check the snapshot id, i would not make an equality check that does copy()

Reply:
https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/snapshot/definition.py#L601

Reply:
I think we need a general solution that doesn't require making sure to override `__eq__` for each equality check.

Reply:
Opening separate PR to try to find a solution everyone is happy with

Reply:
Created here: https://github.com/TobikoData/sqlmesh/pull/2035